pkg/util/mount/mount_linux.go

```diff
@@ -1131,18 +1172,38 @@ func doSafeOpen(pathname string, base string) (int, error) {
 	for _, seg := range segments {
 		currentPath = filepath.Join(currentPath, seg)
 		if !pathWithinBase(currentPath, base) {
-			return -1, fmt.Errorf("path %s is outside of allowed base %s", currentPath, base)
+			return -1, defaultDeviceInfo, fmt.Errorf("path %s is outside of allowed base %s", currentPath, base)
 		}
 
 		glog.V(5).Infof("Opening path %s", currentPath)
 		childFD, err = syscall.Openat(parentFD, seg, nofollowFlags, 0)
 		if err != nil {
-			return -1, fmt.Errorf("cannot open %s: %s", currentPath, err)
+			if err == syscall.ENXIO {
+				var deviceStat unix.Stat_t
+				deviceStatErr := unix.Fstatat(parentFD, seg, &deviceStat, unix.AT_SYMLINK_NOFOLLOW)
+				if deviceStatErr != nil {
+					return -1, defaultDeviceInfo, fmt.Errorf("Error getting stats for %s with %v", currentPath, deviceStatErr)
+				}
+
+				fileFmt := deviceStat.Mode & syscall.S_IFMT
+
+				if fileFmt == syscall.S_IFREG || fileFmt == syscall.S_IFDIR {
```

Comment:
would it be better to check against a whitelist of allowed/expected types? for example, this seems to allow `syscall.S_IFLNK`

Reply:
We can but in odd case where it does happen, what will happen is - inode and device number capture here will belong to symlink itself, whereas bindTarget will resolve the symlink target and have different inode and device number

Reply:
Yeah, prefer whitelisting allowed types and erroring on the rest.  That way we don't accidentally allow something that shouldn't be.